text/1013-datetime-arithmetic.rst

```diff
@@ -0,0 +1,226 @@
+cannot account for DST and so 1 day and 24 hours are always equivalent
+durations in this context.
+
+The ``local_time`` will introduce the following subtraction operator::
+
+  CREATE INFIX OPERATOR
+  std::`-` (
+    l: cal::local_time,
+    r: cal::local_time
+  ) -> cal::relative_duration {
+      USING SQL $$
+          SELECT ("l" - "r")::edgedb.relative_duration_t
+      $$;
+  };
+
+The result is expected to be a ``relative_duration`` in the range [-24 hrs, 24
+hrs].
+
+The ``local_date`` will introduce the following subtraction operator::
+
+  CREATE INFIX OPERATOR
+  std::`-` (
+    l: cal::local_date,
+    r: cal::local_date
+  ) -> int64 {
+      USING SQL $$
+          SELECT ("l" - "r")::int8
+      $$;
+  };
+
+The result is expected to be the number of days.
+
+
+Duration functions
+------------------
+
+We also will introduce normalization functions for ``relative_duration``::
+
+CREATE FUNCTION
+cal::normalize_relative_hours(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_hours';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_days(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_days';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_duration(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_interval';
+};
+
+The ``normalize_relative_hours`` converts 24 hrs chunks into days.
+The ``normalize_relative_days`` converts days into months assuming that 1 month = 30 days.
+The ``normalize_relative_duration`` is the equivalent of calling
+``normalize_relative_days(normalize_relative_hours(x))``.
+
+Notice that for ``relative_duration`` the assumtion that 1 day is always 24
+hours holds true because it represents the time adjustment necessary on some
+clock rather than real time (like ``duration``). Therefore the
+``normalize_relative_hours`` transformation is technically lossless and could
+safely be reversed. However, not every month is 30 days and so the
+``normalize_relative_days`` and ``normalize_relative_duration`` are
+potentially lossy transformations aiming to produce an approximately
+equivalent ``relative_duration`` using months, years, etc. so care must be
+taken when using these conversions.
+
+We also need a function for extracting various ``duration`` and
+``relative_duration`` components::
+
+CREATE FUNCTION
+std::duration_get(dur: cal::relative_duration, el: std::str) -> std::float64
+{
+  ...
+};
+
+CREATE FUNCTION
+std::duration_get(dur: std::duration, el: std::str) -> std::float64
+{
+  ...
+};
+
+The components avaialable for extraction from ``relative_duration`` are:
+*millenium*, *century*, *decade*, *year*, *quarter*, *month*, *day*, *hour*,
+*minutes*, *seconds*, *milliseconds*, *microseconds*, and *totalseconds*. The
+*totalseconds* converts a ``relative_duration`` to seconds represented as a
+``float64`` value.
```

Comment:
We need to document what does `duration_get` for `std::duration`. It never returns anything bigger than `hours`, right? Is asking for that is an error or zero result?

Also why we don't use enum here, is it because of performance issue?

Reply:
Updated the RFC